File: poses_viewer/src/drill2d/__tests__/strokeCycle2d.test.ts
import { describe, expect, it } from 'vitest'
import { makeCycle, Stroke2D } from '../types'

/** Mirrors StrokeCycle2DTest.kt. */
const stroke = (
  i: number,
  start: number,
  peak: number,
  end: number,
  speed = 2.4,
): Stroke2D => ({
  strokeIndex: i,
  startFrame: start,
  peakFrame: peak,
  endFrame: end,
  peakSpeed: speed,
})

describe('StrokeCycle2D (mirrors StrokeCycle2DTest)', () => {
  it('paired cycle forwards fields from drive', () => {
    const backswing = stroke(0, 10, 20, 30)
    const drive = stroke(1, 30, 45, 60, 3.1)
    const cycle = makeCycle(backswing, drive)

    expect(cycle.peakFrame).toBe(drive.peakFrame)
    expect(cycle.peakSpeed).toBe(drive.peakSpeed)
    expect(cycle.startFrame).toBe(backswing.startFrame)
    expect(cycle.endFrame).toBe(drive.endFrame)
  })

  it('unpaired cycle uses drive startFrame', () => {
    const drive = stroke(0, 30, 45, 60, 2.8)
    const cycle = makeCycle(null, drive)

    expect(cycle.backswing).toBeNull()
    expect(cycle.peakFrame).toBe(drive.peakFrame)
    expect(cycle.peakSpeed).toBe(drive.peakSpeed)
    expect(cycle.startFrame).toBe(drive.startFrame)
    expect(cycle.endFrame).toBe(drive.endFrame)
  })
})

File: shared/src/commonTest/kotlin/com/ttcoachai/shared/models/StrokeCycle2DTest.kt
package com.ttcoachai.shared.models

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class StrokeCycle2DTest {

    private fun stroke(i: Int, start: Int, peak: Int, end: Int, speed: Float = 2.4f) = Stroke2D(
        strokeIndex = i,
        startFrame = start,
        peakFrame = peak,
        endFrame = end,
        peakSpeed = speed
    )

    @Test
    fun pairedCycleForwardsFieldsFromDrive() {
        val backswing = stroke(0, start = 10, peak = 20, end = 30)
        val drive = stroke(1, start = 30, peak = 45, end = 60, speed = 3.1f)
        val cycle = StrokeCycle2D.of(backswing, drive)

        assertEquals(drive.peakFrame, cycle.peakFrame)
        assertEquals(drive.peakSpeed, cycle.peakSpeed)
        assertEquals(backswing.startFrame, cycle.startFrame)
        assertEquals(drive.endFrame, cycle.endFrame)
    }

    @Test
    fun unpairedCycleUsesDriveStartFrame() {
        val drive = stroke(0, start = 30, peak = 45, end = 60, speed = 2.8f)
        val cycle = StrokeCycle2D.of(null, drive)

        assertNull(cycle.backswing)
        assertEquals(drive.peakFrame, cycle.peakFrame)
        assertEquals(drive.peakSpeed, cycle.peakSpeed)
        assertEquals(drive.startFrame, cycle.startFrame)
        assertEquals(drive.endFrame, cycle.endFrame)
    }
}
